packages/block-editor/src/components/block-list/index.js

```diff
@@ -43,71 +35,94 @@ const forceSyncUpdates = ( WrappedComponent ) => ( props ) => {
-		// relative to the document, so fire it with the last known position.
-		this.onScroll = () => this.onPointerMove( { clientY: this.lastClientY } );
-
-		this.lastClientY = 0;
-		this.nodes = {};
-	}
-
-	componentDidMount() {
-		window.addEventListener( 'mousemove', this.setLastClientY );
-	}
+		this.setSelection = this.setSelection.bind( this );
 
-	componentWillUnmount() {
-		window.removeEventListener( 'mousemove', this.setLastClientY );
-	}
-
-	setLastClientY( { clientY } ) {
-		this.lastClientY = clientY;
-	}
-
-	setBlockRef( node, clientId ) {
-		if ( node === null ) {
-			delete this.nodes[ clientId ];
-		} else {
-			this.nodes = {
-				...this.nodes,
-				[ clientId ]: node,
-			};
-		}
+		this.ref = createRef();
 	}
 
 	/**
-	 * Handles a pointer move event to update the extent of the current cursor
-	 * multi-selection.
-	 *
-	 * @param {MouseEvent} event A mousemove event object.
+	 * When the component updates, and there is multi selection, we need to
+	 * select the entire block contents.
 	 */
-	onPointerMove( { clientY } ) {
-		// We don't start multi-selection until the mouse starts moving, so as
-		// to avoid dispatching multi-selection actions on an in-place click.
-		if ( ! this.props.isMultiSelecting ) {
-			this.props.onStartMultiSelect();
+	componentDidUpdate() {
+		const {
+			hasMultiSelection,
+			blockClientIds,
+			// These must be in the right DOM order.
+			multiSelectedBlockClientIds,
+		} = this.props;
+
+		if ( ! hasMultiSelection ) {
+			return;
 		}
 
-		const blockContentBoundaries = getBlockDOMNode( this.selectionAtStart ).getBoundingClientRect();
+		const { length } = multiSelectedBlockClientIds;
+		const start = multiSelectedBlockClientIds[ 0 ];
+		const end = multiSelectedBlockClientIds[ length - 1 ];
+		const startIndex = blockClientIds.indexOf( start );
 
-		// prevent multi-selection from triggering when the selected block is a float
-		// and the cursor is still between the top and the bottom of the block.
-		if ( clientY >= blockContentBoundaries.top && clientY <= blockContentBoundaries.bottom ) {
+		// The selected block is not in this block list.
+		if ( startIndex === -1 ) {
 			return;
 		}
 
-		const y = clientY - blockContentBoundaries.top;
-		const key = findLast( this.coordMapKeys, ( coordY ) => coordY < y );
+		let startNode = this.ref.current.querySelector(
+			`[data-block="${ start }"]`
+		);
+		let endNode = this.ref.current.querySelector(
+			`[data-block="${ end }"]`
+		);
+
+		const selection = window.getSelection();
+		const range = document.createRange();
 
-		this.onSelectionChange( this.coordMap[ key ] );
+		// The most stable way to select the whole block contents is to start
+		// and end at the deepest points.
+		startNode = getDeepestNode( startNode, 'start' );
+		endNode = getDeepestNode( endNode, 'end' );
+
+		range.setStartBefore( startNode );
+		range.setEndAfter( endNode );
+
+		selection.removeAllRanges();
+		selection.addRange( range );
```

Comment:
This logic here is probably better extracted to its own function as `componentDidUpdate` could potentially do more than just multi-selection range syncing (if I understand properly what is this doing :) )

Reply:
I feel rewriting this component with hooks could make it more readable but that can be a separate  PR

Reply:
Right, this could be improved for sure.

Reply:
Do you feel like I should extract it now for this PR, or rewrite it later as hooks?

Reply:
I've extracted it now.